### PR TITLE
Rework announcements

### PR DESCRIPTION
**Pull request by daun**
## Motivation

- Prompted by #41
- Been doing some research on the current state of the art regarding announcing new content after a fetch
- The consensus seems to be a permanent `assertive` and `atomic` live region without a specific `role`

## Description

- Create a custom announcer element, i.e. live region
  - Using `assertive` without a `role` seems to solve the issue in #41
  - Get rid of current third-party dependency
- Focus and announce new content in `visit:end`
- Make `content:announce` and `content:focus` synchronous hooks
- Tested to work in VoiceOver, NVDA & JAWS

## Checks

- [x] The PR is submitted to the `master` branch
- [x] The code was linted before pushing (`npm run lint`)
- [ ] ~~All tests are passing (`npm run test`)~~
- [ ] ~~New or updated tests are included~~
- [ ] ~~The documentation was updated as required~~

## Prior art

Some examples how this problem is solved in other client-side routers and frameworks.

### [SvelteKit](https://github.com/sveltejs/kit/blob/38294bef6ae347b2153aea9f9f56bed025a0cebd/packages/kit/src/core/sync/write_root.js#L114)

Assertive, atomic, no role, permanent, timeout of `0`.

```html
<div aria-live="assertive" aria-atomic="true">{title}</div>
```

### [Next](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/route-announcer.tsx#L54C5-L61C9)

Assertive, non-atomic, alert, permanent, timeout of `0`.

```html
<p aria-live="assertive" role="alert">{route}</p>
```

### [Astro](https://github.com/withastro/astro/blob/fd18a5dda981c1b13c0a585d174ed2f09a2e9218/packages/astro/src/transitions/router.ts#L50)

Assertive, atomic, no role, dynamically created, timeout of `60`.

```html
<div aria-live="assertive" aria-atomic="true">{title}</div>
```

```js
// Much thought went into this magic number; screen readers need to see
// that the element changed and might not do so if it happens too quickly.
setTimeout(() => {
  div.textContent = title
}, 60)
```

### [Ember](https://github.com/ember-a11y/a11y-announcer/blob/master/app/initializers/add-announcer-to-router.js)

Polite, non-atomic, no role, permanent, timeout of `100`.

```js
this._timerId = later(() => {
  this.get('announcer').announce(message, 'polite')
}, 100)

### CODE REVIEW

**Comment by hirasso** [comment]
That's some amazing research!! 👏

**Comment by daun** [comment]
@hirasso I've been testing this in VoiceOver, as well as NVDA und (finally) JAWS on a virtual machine. All working great so far. Ready to merge from my end. Let me know if you want to give it a spin yourself, but if I remember correctly you've been testing this branch in a project of yours, right?

**Comment by hirasso** [comment]
@daun I was so busy the last few days that I didn't yet come around to installing the branch on my current project. Would it be okay for you if I do this now and report back to you?

**Comment by daun** [comment]
@hirasso Sure, take your time! There's no rush and I prefer getting this right. Let me know if you need pointers about getting Jaws to run, but it's basically just a Parallels install of Windows with a Jaws test license running for 40 minutes at a time 🤠

**Review by hirasso — CHANGES_REQUESTED** [review]
Works great! One issue I came across while testing: If the heading element is being morphed by swup-morph-plugin, the `prepareAnnouncement` function is being executed too early, announcing the previous page title. My HTML:

`<h1 data-swup-morph="page-title">My Title</h1>`

To fix this we could either

- move the announcement logic to `visit:end`
- use a high number (run later) for the hook. This, for example, worked in my test:

```js
this.on('content:replace', this.prepareAnnouncement, { priority: 100 });
```

**Comment by hirasso** [comment]
@daun in my test, `handleNewPageContent` keeps working even without the `priority` setting – I guess due to `nextTick()` in that function? Anyways – I actually prefer `visit:end` as the point in time where the announcement happens. At that stage we can be completely sure that everything new has landed in the document and the visit is still valid: 

- `visit:end` will only being executed if the visit wasn't aborted in the meantime
- In my test, the screen reader is very busy reading all kinds of other stuff anyways before it announces the page visit. It happens long after `visit:end` 
- Also, intuitively it makes sense to me if the plugin announces the new page right after it removes `[aria-busy=true]` from the `documentElement`

I'm also a bit worried about the use of the `priority` option in plugins hooks: the magic number ( `100` in my test ) doesn't mean anything and any other code (plugin or consumer) could just hook in later in the queue, without being aware the implications.

**Comment by hirasso** [comment]
actually... can you remember what's the reason for using `nextTick` in `handleNewPageContent`?

**Comment by hirasso** [comment]
Hooking into `visit:end` brings other challenges. The two hooks `content:announce` and `content:focus` won't be executed since the visit was already marked as `COMPLETED`. I remember there was an approach we had to situations like these? Can you remember what it was?

**Comment by hirasso** [comment]
> actually... can you remember what's the reason for using `nextTick` in `handleNewPageContent`?

Ah, now I think I get it. `nextTick` is there to make sure everything has settled. Basically like a more robust alternative to the `priority` setting? So if we would end up deciding to keep using `content:replace`, we should move `prepareAnnouncent` in the nextTick closure.

**Comment by daun** [comment]
@hirasso Right, the wait for the next tick was there to let any potential dom updates from user hook handlers settle before reading out or focussing the new content. Moving everything to visit:end should work, if we remove the nextTick call. There will be a potential conflict with the autofocus part, which kicks in on visit:end. Will need to juggle this a bit before having an opinion 🍊

**Comment by hirasso** [comment]
> There will be a potential conflict with the autofocus part, which kicks in on visit:end.

Ah that makes sense! Sure, take your time investigating.

**Comment by daun** [comment]
It's actually much cleaner now :) I've moved everything into `visit:end`, removed the next-tick wait, and collapsed the body focus and autofocus into a single handler. Testing on Windows right now.

**Comment by daun** [comment]
I was tempted to refactor the whole thing in this PR, but I must stay strong (🤡) until this is merged and we've solved the anchor-link focus issue as well. Then I'd like to make everything more testable and actually write the tests before the release...

**Comment by daun** [comment]
Works great in NVDA und JAWS as well.

File: src/index.ts [hunk 1]
```diff
@@ -145,10 +138,35 @@ export default class SwupA11yPlugin extends Plugin {
...
+
+			// Found and focused [autofocus] element? Return early
+			if (this.focusAutofocusElement() === true) return;
+
+			// Otherwise, find and focus actual content container
+			this.focusContentElement(visit.a11y.focus);
+		});
+	}
+
+	getDocumentTitle(): string | undefined {
```
Comment (hirasso) [inline src/index.ts:169]: I find `getDocumentTitle` a bit misleading. That term usually means the `<title>` element of the document for me. How about `getPageTitle` or `getPageAnnouncement`?

File: src/index.ts [hunk 2]
```diff
@@ -158,89 +176,52 @@ export default class SwupA11yPlugin extends Plugin {
...
-	getAutofocusElement(): HTMLElement | undefined {
-		const focusEl = document.querySelector<HTMLElement>('body [autofocus]');
-		if (focusEl && !focusEl.closest('inert, [aria-disabled], [aria-hidden="true"]')) {
-			return focusEl;
+	focusAutofocusElement(): boolean {
+		if (!this.options.autofocus) return false;
+
+		const autofocusEl = getAutofocusElement();
+		if (autofocusEl && autofocusEl !== document.activeElement) {
+			if (autofocusEl !== document.activeElement) {
```
Comment (hirasso) [inline src/index.ts:219]: This looks redundant to me. Also nitpick: I like to use `isSameNode` – but `===` is just as good.

File: src/index.ts [hunk 3]
```diff
@@ -158,89 +176,52 @@ export default class SwupA11yPlugin extends Plugin {
...
+				autofocusEl.focus(); // no preventScroll flag here, as probably intended
+			}
+			return true;
+		} else {
```
Comment (hirasso) [inline src/index.ts:223]: Can we get rid of the unnecessary `else` here:

```js
if (condition) {
  //...
  return true;
}
return false;
```

**Review by hirasso — CHANGES_REQUESTED** [review]

**Comment by hirasso** [comment]
Works great now with the hooks on `visit:end`!

**Comment by daun** [comment]
@hirasso Ready to go :)

**Review by hirasso — APPROVED** [review]

File: src/index.ts [hunk 4]
```diff
@@ -158,90 +176,53 @@ export default class SwupA11yPlugin extends Plugin {
...
-	getAutofocusElement(): HTMLElement | undefined {
-		const focusEl = document.querySelector<HTMLElement>('body [autofocus]');
-		if (focusEl && !focusEl.closest('inert, [aria-disabled], [aria-hidden="true"]')) {
-			return focusEl;
+	focusAutofocusElement(): boolean {
+		if (!this.options.autofocus) return false;
+
+		const autofocusEl = getAutofocusElement();
+		if (autofocusEl) {
+			if (autofocusEl !== document.activeElement) {
```
Comment (hirasso) [inline src/index.ts:219]: I'd prefer removing the nested if, in favour of the `&&`. But that's a matter of taste.

**Review by hirasso — COMMENTED** [review]
One last little nitpick

File: src/index.ts (hunk 4 above)
Comment (daun) [reply src/index.ts:219]: The `&&` was a bug before, it's actually two separate conditions. It needs to return `true` if the element was found, but only execute `.focus()` if it isn't the active element. The alternative would by `else` + `elseif` + `else`, which isn't more readable in my opinion.

Comment (hirasso) [reply src/index.ts:219]: Aah! Interesting. We need more code docs :)

**Comment by hirasso** [comment]
Ready to go from my end 🛫

File: src/index.ts (hunk 4 above)
Comment (daun) [reply src/index.ts:219]: Yeah, those will be part of the refactor 😅

**Comment by daun** [comment]
Yeehaw 🤠 Probably my favorite PR so far. A bit of research, a lot of trial and error, outside contributors, very nice :)

**Pull request merged**